include/envoy/thread/thread.h

```diff
@@ -2,40 +2,149 @@
 
 #include "envoy/common/pure.h"
 
+#include "absl/base/thread_annotations.h"
+
 namespace Envoy {
 namespace Thread {
 
 /**
- * Like the C++11 "basic lockable concept" but a concrete interface vs. a template.
+ * Like the C++11 "basic lockable concept" but a pure virtual interface vs. a template, and
+ * with thread annotations.
  */
-class BasicLockable {
+class LOCKABLE BasicLockable {
 public:
   virtual ~BasicLockable() {}
 
-  virtual void lock() PURE;
-  virtual bool try_lock() PURE;
-  virtual void unlock() PURE;
+  virtual void lock() EXCLUSIVE_LOCK_FUNCTION() PURE;
+  virtual bool tryLock() EXCLUSIVE_TRYLOCK_FUNCTION(true) PURE;
+  virtual void unlock() UNLOCK_FUNCTION() PURE;
 };
 
 /**
  * A lock guard that deals with an optional lock.
  */
-template <class T> class OptionalLockGuard {
+class SCOPED_LOCKABLE OptionalLockGuard {
+public:
+  /**
+   * Establishes a scoped mutex-lock. If non-null, the mutex is locked upon construction.
+   *
+   * @param lock the mutex.
+   */
+  OptionalLockGuard(BasicLockable* lock) EXCLUSIVE_LOCK_FUNCTION(lock) : lock_(lock) {
+    if (lock_ != nullptr) {
+      lock_->lock();
+    }
+  }
+
+  /**
+   * Destruction of the OptionalLockGuard unlocks the lock, if it is non-null.
+   */
+  ~OptionalLockGuard() UNLOCK_FUNCTION() {
+    if (lock_ != nullptr) {
+      lock_->unlock();
+    }
+  }
+
+private:
+  BasicLockable* lock_;
+};
+
+/**
+ * Like LockGuard, but uses a tryLock() on construction rather than a lock().
+ */
+class SCOPED_LOCKABLE TryLockGuard {
 public:
-  OptionalLockGuard(T* lock) : lock_(lock) {
-    if (lock) {
-      lock->lock();
+  /**
+   * Establishes a scoped mutex-lock; the a mutex lock is attempted via tryLock, so
+   * an expected outcome is that the lock may fail. isLocked() must be called to
+   * determine whether he lock was actually acquired.
+   *
+   * @param lock the mutex.
+   */
+  TryLockGuard(BasicLockable& lock) EXCLUSIVE_TRYLOCK_FUNCTION(true)
```

Comment:
This doesn't have a return value, so it seems wrong to annotate as a trylock function. But I don't know how to actually annotate the class correctly.

Reply:
Yeah, this is really an abstraction in effect anyway, so might as well use semantics appropriate for a return-less constructor.